mulog-prometheus/src/com/brunobonacci/mulog/publishers/prometheus/metrics.clj

```diff
@@ -0,0 +1,81 @@
+(ns com.brunobonacci.mulog.publishers.prometheus.metrics
+  (:require [clojure.string :as str]
+            [com.brunobonacci.mulog.publishers.prometheus.metrics-spec :refer [invalid-metric-name-chars
+                                                                               invalid-metric-label-chars
+                                                                               reserved-metric-label-chars]]))
+
+(defn- exception?
+  [e]
+  (instance? java.lang.Exception e))
+
+
+
+(defn- cleanup-event
+  [event]
+  (->> event
+    (#(dissoc % :mulog/trace-id :mulog/parent-trace :mulog/root-trace :mulog/timestamp))
+    (filter (fn [[_ v]] (or (string? v) (number? v) (keyword? v) (exception? v))))
+    (into {})))
+
+
+
+(defn- kw-str
+  [k]
+  (-> (if (keyword? k)
+        (if (namespace k)
+          (str (namespace k) "_" (name k))
+          (name k))
+        (str k))
+    (str/replace invalid-metric-name-chars "_")))
+
+
+
+(defn- as-labels
+  [m]
+  (->> m
+    (#(dissoc % :mulog/namespace :mulog/event-name))
+    (map (fn [[k v]]
+           [(->
+              (cond
+                (keyword? k) (kw-str k)
+                :else (str k))
+              (str/replace invalid-metric-label-chars  "_")
+              (str/replace reserved-metric-label-chars "_"))
+            (cond
+              (keyword? v) (name v)
+              (exception? v) (str (type v) ": " (.getMessage ^Exception v))
+              :else (str v))]))
+    (into {})))
+
+
+
+(defn- event->metrics
+  [{:keys [mulog/namespace mulog/event-name] :as event}]
+  (let [numeric      (filter (comp number? second) event)
+        namespace    (kw-str namespace)
+        event-name   (kw-str event-name)
+        labels       (as-labels event)]
+    (conj (for [type  [:gauge :histogram :summary]
+                [k v] numeric
+                :let  [key-str (kw-str k)
+                       e-name  (str event-name "_" key-str)]]
```

Comment:
I don't understand this. I don't think we should publish every metric 3 times in all possible formats.

The approach i suggested was:
  - `mulog/duration` -> is exposed as a summary
  - all other numerical values are exposed as gauges

so every metrics is exposed only once. if the uses decide for histograms instead of summary he can do that via `transform-metrics`.

Reply:
Sorry about this. I misunderstood the initial conversation. This would explain why the name `timer_nanos` was originally there and I didn't need to create a unique metric name for every type.

I have made the above correction